### DISCUSSION TITLE
Discount should be applied before tax is caclulated

### DISCUSSION THREAD

**Discussion by yhortuk** [lunarphp/lunar#1046] [Feature Requests]
- Lunar version: v0.2-rc1

### Expected Behaviour:

In cart:
Item 1.96 x 2 = 3.92 USD
Shipping = 13  USD
Discount 10% = 0.39
**Vat 24% = (3.92 - 0.39 + 13) * 0.24 = 3.97**

### Actual Behaviour:

In cart:
Item 1.96 x 2 = 3.92 USD
Shipping = 13  USD
Discount 10% = 0.39
**Vat 24% = (3.92 + 13) * 0.24 = 4.06**

### Steps To Reproduce:

1) Add tax rule
2) Create discount
3) Create a cart instance
4) Add item and shipping option to the cart
5) Call $cart->calculate()
6) Call $cart->taxTotal

**Comment by glennjacobs** [comment]
I understand the problem. 

Maybe we need the option to specify whether the discount is before or after tax. 

Eg. In the UK, retail sales always show the price including tax. So if the item was £50.00, the customer would expect a 20% discount to make it £40.00, not £38.40.

**Comment by glennjacobs** [comment]
Following up on this, I consulted accounting software. You are calculating the tax from the cart sub-total in your example, whereas tax should be calculated per line.

E.g. 

```
Product $1.96 x 2 = $3.92
Discount 10% applied before tax = $0.39 (rounded to 2 decimal places)
Discounted amount = $3.53
Tax @ 24% = 0.85 (rounded to 2 decimal places)
Line total inc. tax = $4.38
```

We are going to be updating the discount code soon and will introduce tests like this to ensure calculations are correct.
